tf2_ros/include/tf2_ros/message_filter.h

```diff
@@ -429,7 +429,8 @@ class MessageFilter : public MessageFilterBase, public message_filters::SimpleFi
         --message_count_;
       }
 
-      // Add the message to our list
+      // Add the message to our list while keeping a lock on the messages.
+      std::unique_lock<std::mutex> unique_lock(messages_mutex_);
```

Comment:
Instead of adding this here, just move this to the beginning of the else block (https://github.com/ros2/geometry2/pull/279/files#diff-724ff79b0edd5a83f8a62b5e2ffa085cR412), then https://github.com/ros2/geometry2/pull/279/files#diff-724ff79b0edd5a83f8a62b5e2ffa085cR416 .  That should cover both cases.